zcash_primitives/src/transaction/components.rs

```diff
@@ -19,7 +19,7 @@ const PHGR_PROOF_SIZE: usize = (33 + 33 + 65 + 33 + 33 + 33 + 33 + 33);
 const ZC_NUM_JS_INPUTS: usize = 2;
 const ZC_NUM_JS_OUTPUTS: usize = 2;
 
-#[derive(Debug)]
+#[derive(Clone, Debug, PartialEq, Eq)]
```

Comment:
Do you actually require `Eq` to be implemented? I know that `PartialEq` is usually required for test comparisons.

Reply:
The reason to include `Eq` (which is derived from `PartialEq`) is to enforce `x == x` (*reflexivity*), which *is* relevant for this type.

`PartialEq`'s are not strict *[equivalence relations](https://en.wikipedia.org/wiki/Equivalence_relation)* because they do not enforce `x == x`.

This counterintuitive property makes sense when you think about the concrete example of IEEE floating-point values.   They are specified to be `NaN` in some cases (e.g. `0.0/0.0`), but `NaN != NaN`, i.e. `NaN` is _not_ reflexive.

e.g.:
```Rust
assert_eq!(0.0/0.0 == 0.0/0.0, false)
```

Since `f32` and `f64` are the **only** types in the standard library that are `PartialEq`, but not `Eq`, it's generally best practice to implement `Eq` on any type that doesn't include them.

In particular this provides the type with the intuitive meaning of:

`==`

Most of what I am asserting here is rehashed from the `Equality Tests` section of this book:

http://shop.oreilly.com/product/0636920040385.do

Reply:
Here's a lot of the same info, but readable!!
https://doc.rust-lang.org/std/cmp/trait.PartialEq.html

Reply:
See also:  https://www.reddit.com/r/rust/comments/3f9e7q/why_does_eq_require_partialeq/

Reply:
NOTE:  The reason this type can be used in equivalences without explicitly defining "`fn eq`" in either the `PartialEq`, or `Eq` trait (either as a default, or when implementing for the type), is because the `PartialEq` trait is being implemented with the `#[derive(` macro which is supplying an `fn eq` implementation.

At least...   I think that's what's happening, when I drilled into the macro implementation I hit this:

https://doc.rust-lang.org/src/core/cmp.rs.html#209

When I replace the `#[derive(PartialEq)]` with an explicit:

`impl PartialEq for OutPoint {}` I get the compiler error I expect:

```Rust
error[E0046]: not all trait items implemented, missing: `eq`
  --> src/main.rs:11:1
   |
11 | impl PartialEq for OutPoint {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `eq` in implementation
   |
   = note: `eq` from trait: `fn(&Self, &Rhs) -> bool`
```

Reply:
This is consistent with this documentation (and the fact that `u8`'s are `Eq`):
https://doc.rust-lang.org/std/cmp/trait.PartialEq.html#derivable

Reply:
TL;DR   `OutPoint`s should be `Eq`

Reply:
Technically under the [BIP 30](https://github.com/bitcoin/bips/blob/master/bip-0030.mediawiki) consensus rules we inherited from Bitcoin Core, `OutPoints` do not have full equivalence, because transaction IDs are allowed to repeat in the block chain as long as they do not repeat in the UTXO set. Thus an `OutPoint` is a unique reference in the context of a block chain (to the transaction with that ID in the most recent block), but not technically unique on its own (just treated that way).

In practice, wallet code assumes that transaction IDs will never repeat (which causes the overwriting problem that BIP 30 was written to solve). So maybe we're fine assuming that we can treat identical `OutPoint`s as equivalent.